### PR TITLE
Scheme's Build step - Pre and Post actions prototype

### PR DESCRIPTION
**Pull request by chiragramani**
**Description**

This PR attempts to add Pre and Post actions for the Scheme's Build Step. We generally use this for logging some build analytics/tracking and hence this becomes important for such use-cases.

- Pre and post actions are supported in all the steps of the Scheme steps ("Build", "Run", "Test", "Debug", "Analyze", "Archive"). Though this diff only adds support for the "Build" step and once this prototype is approved, we can add the same support to other steps as well, if required.
- The PR extends the existing build action to also take in the actions(apart from the targets) as the actions are tied to the Build step of the scheme and are not a separate entity.
- I have currently tested this for both BwX and BwB modes and things work as expected. 

I would appreciate some help with the following:

- I don't feel confident about the intersection of this PR with the focussed and unfocussed targets logic in xcode_schemes.bzl. Though it works but not sure if there are other edge cases involved where the current logic might not work. 
- TODOs: once the above is in good shape, I can update an existing example or add a new one with the required tests.
- Please feel free to point out if maybe we should address this support differently.

### CODE REVIEW

File: xcodeproj/internal/xcode_schemes.bzl [hunk 1]
```diff
@@ -157,7 +177,7 @@ def make_xcode_schemes(bazel_labels):
         A `struct` that can be used as a `bazel_labels` module.
     """
 
-    def _build_action(targets):
+    def _build_action(targets, pre_actions = [], post_actions = []):
```
Comment (brentleyjones) [inline xcodeproj/internal/xcode_schemes.bzl:180]:
```suggestion
    def _build_action(targets, *, pre_actions = [], post_actions = []):
```

We should require the labels once the call gets this complex.

File: tools/generator/src/DTO/XcodeScheme.swift [hunk 2]
```diff
@@ -101,7 +101,9 @@ disabled, but the target is referenced in the scheme's \(keyPath.actionType) act
 
             // Create a new build action which includes all of the referenced labels as build targets
             // We must do this after processing all of the other actions.
-            let newBuildAction = try XcodeScheme.BuildAction(targets: buildTargets.values)
+            let newBuildAction = try XcodeScheme.BuildAction(targets: buildTargets.values,
+                                                             preActions: buildAction?.preActions ?? [],
+                                                             postActions: buildAction?.postActions ?? [])
```
Comment (brentleyjones) [inline tools/generator/src/DTO/XcodeScheme.swift:106]: Style nit:
```suggestion
            let newBuildAction = try XcodeScheme.BuildAction(
                targets: buildTargets.values,
                preActions: buildAction?.preActions ?? [],
                postActions: buildAction?.postActions ?? []
            )
```

File: tools/generator/src/DTO/XcodeScheme.swift [hunk 3]
```diff
@@ -132,11 +134,21 @@ extension XcodeScheme {
 }
 
 extension XcodeScheme {
+    struct PrePostAction: Equatable, Decodable {
+        let name: String
+        let target: BazelLabel
+        let scriptContents: String
+    }
+
```
Comment (brentleyjones) [inline tools/generator/src/DTO/XcodeScheme.swift:142]: Since this will apply to more than just the `BuildAction` in the future, can you extract it to it's own `extension` above this one?

File: tools/generator/src/Extensions/XcodeScheme+Extensions.swift [hunk 4]
```diff
@@ -168,7 +168,7 @@ extension XcodeScheme {
     var allBazelLabels: Set<BazelLabel> {
         var labels = Set<BazelLabel>()
         if let buildAction = buildAction {
-            labels.formUnion(buildAction.targets.map(\.label))
+            labels.formUnion(buildAction.targets.map(\.label) +  buildAction.preActions.map(\.target) + buildAction.postActions.map(\.target))
```
Comment (brentleyjones) [inline tools/generator/src/Extensions/XcodeScheme+Extensions.swift:171]:
```suggestion
            labels.formUnion(buildAction.targets.map(\.label))
            labels.formUnion(buildAction.preActions.map(\.target))
            labels.formUnion(buildAction.postActions.map(\.target))
```

File: xcodeproj/internal/xcode_schemes.bzl [hunk 5]
```diff
@@ -147,6 +160,13 @@ def _unfocus_schemes(schemes, unfocused_targets):
 
     return focused_schemes
 
+def _pre_post_action(name, target, script_contents):
+    return struct(
+        name = name,
+        target = target,
```
Comment (brentleyjones) [inline xcodeproj/internal/xcode_schemes.bzl:166]: This can be `None`, right? Pre/post actions don't require (though usually use) an `environmentBuildable`.

File: xcodeproj/internal/xcode_schemes.bzl [hunk 6]
```diff
@@ -147,6 +160,13 @@ def _unfocus_schemes(schemes, unfocused_targets):
 
     return focused_schemes
 
+def _pre_post_action(name, target, script_contents):
```
Comment (brentleyjones) [inline xcodeproj/internal/xcode_schemes.bzl:163]: We should ensure that the only thing required is `script_contents`. `name` can default to `"Run Script"`: https://github.com/tuist/XcodeProj/blob/c36d3a8501588e280f4199b00c5cfc5a23b27dbf/Sources/XcodeProj/Scheme/XCScheme%2BExecutionAction.swift#L16

File: tools/generator/xcodeproj_targets.bzl [hunk 7]
```diff
@@ -37,12 +37,24 @@ def get_xcode_schemes():
...
+                targets = [
+                    xcode_schemes.build_target(
+                        _APP_TARGET,
+                        xcode_schemes.build_for(archiving = True),
+                    ),
+                ],
+                pre_actions = [
+                    xcode_schemes.pre_action("Example: Start build time tracking...",
+                    _APP_TARGET,
+                    "echo 'Building target: generator'"),
```
Comment (brentleyjones) [inline tools/generator/xcodeproj_targets.bzl:50]:
```suggestion
                    xcode_schemes.pre_action(
                        name = "Example: Start build time tracking...",
                        target = _APP_TARGET,
                        script = "echo 'Building target: generator'",
                    ),
```

File: xcodeproj/internal/xcode_schemes.bzl (hunk 6 above)
Comment (brentleyjones) [inline xcodeproj/internal/xcode_schemes.bzl:163]:
```suggestion
def _pre_post_action(name, target, script):
```

**Review by brentleyjones — COMMENTED** [review]
I think this is the right direction. Thank you @chiragramani!

I would love @cgrindel's feedback as well, as he did most of the scheme work before now.

I left various random comments. Some of it is to clean it up to match the existing codebase (you can run swiftformat and swiftlint with the repo settings to get most of the way there).

File: xcodeproj/internal/xcode_schemes.bzl [hunk 8]
```diff
@@ -34,6 +43,8 @@ def _focus_schemes(schemes, focused_targets):
             if build_targets:
                 build_action = xcode_schemes_internal.build_action(
                     targets = build_targets,
+                    pre_actions = _focus_actions(build_targets, build_action.pre_actions),
```
Comment (brentleyjones) [inline xcodeproj/internal/xcode_schemes.bzl:46]: Since the pre/post actions won't need build targets (we can have a `build_action` with just pre/post actions, and they might not even have a `target` set) this won't work correctly. 

I'll post up a fixed version to the PR before we merge (so you can ignore this, this is a comment for me).

File: xcodeproj/internal/xcode_schemes.bzl (hunk 6 above)
Comment (brentleyjones) [reply xcodeproj/internal/xcode_schemes.bzl:163]: With that said, we should require `target` (don't set a default), but allow `None`. This at least encourages people to set it, as it's most likely needed for environment variables to be as expected.

File: xcodeproj/internal/xcode_schemes.bzl [hunk 9]
```diff
@@ -6,6 +6,15 @@ load(":xcode_schemes_internal.bzl", "xcode_schemes_internal")
...
 _DEFAULT_BUILD_CONFIGURATION_NAME = "Debug"
 
+
+def _focus_actions(build_targets, actions):
+    focussed_actions = []
+    build_target_labels = [target.label for target in build_targets]
+    for action in actions:
+        if action.target in build_target_labels:
+            focussed_actions.append(action)
+    return focussed_actions
```
Comment (erikkerber) [inline xcodeproj/internal/xcode_schemes.bzl:16]: Super nitty, but the admittedly US-centric `focused` spelling is used elsewhere in this repo.

```suggestion
    focused_actions = []
    build_target_labels = [target.label for target in build_targets]
    for action in actions:
        if action.target in build_target_labels:
            focused_actions.append(action)
    return focused_actions
```

File: xcodeproj/internal/xcode_schemes.bzl (hunk 6 above)
Comment (erikkerber) [reply xcodeproj/internal/xcode_schemes.bzl:163]: I was going to leave a comment on the lack of Shell option, but it looks like from that link it's not currently supported by Tuist?

**Review by erikkerber — COMMENTED** [review]
Awesome and thank you so much 🙏🏼 

Agree with Brentley, and defer until @cgrindel chimes in

File: xcodeproj/internal/xcode_schemes.bzl (hunk 6 above)
Comment (brentleyjones) [reply xcodeproj/internal/xcode_schemes.bzl:163]: @maxwellE Is that something you can add to `XcodeProj`?

Comment (maxwellE) [reply xcodeproj/internal/xcode_schemes.bzl:163]: Yes I can look into that

Comment (brentleyjones) [reply xcodeproj/internal/xcode_schemes.bzl:163]: Seems someone was thinking of the same thing? https://github.com/tuist/XcodeProj/pull/721#pullrequestreview-1098649282

File: tools/generator/src/Generator/XCSchemeInfo+BuildActionInfo.swift [hunk 10]
```diff
@@ -29,16 +40,29 @@ extension XCSchemeInfo.BuildActionInfo {
...
-                    buildFor: buildTarget.buildFor
-                )
-            }
+                    .init(
+                        targetInfo: .init(
+                            resolveHostFor: buildTarget.targetInfo,
+                            topLevelTargetInfos: topLevelTargetInfos
+                        ),
+                        buildFor: buildTarget.buildFor
+                    )
```
Comment (cgrindel) [inline tools/generator/src/Generator/XCSchemeInfo+BuildActionInfo.swift:62]: nit: This indent seems incorrect. Can you check this?

File: tools/generator/src/Generator/XCSchemeInfo+BuildActionInfo.swift [hunk 11]
```diff
@@ -65,6 +89,27 @@ extension XCSchemeInfo.BuildActionInfo {
...
+            try actions.map { action in
+                let targetID = try targetIDsByLabel.value(
+                    for: action.target,
+                    context: "creating a `BuildActionInfo`"
+                )
+                return .init(name: action.name,
+                             target: try targetResolver.targetInfo(targetID: targetID),
+                             scriptContents: action.scriptContents)
+            }
+        }
```
Comment (cgrindel) [inline tools/generator/src/Generator/XCSchemeInfo+BuildActionInfo.swift:102]: I suggest moving this to a [custom init like this one in BuildActionInfo](https://github.com/buildbuddy-io/rules_xcodeproj/blob/main/tools/generator/src/Generator/XCSchemeInfo%2BBuildActionInfo.swift#L49-L53).

File: xcodeproj/internal/xcode_schemes.bzl (hunk 6 above)
Comment (brentleyjones) [reply xcodeproj/internal/xcode_schemes.bzl:163]: Now that https://github.com/buildbuddy-io/rules_xcodeproj/commit/7e2fab47ceebbce92771ff17e7a0011efa491eab is merged, we can add `shell` here as well.

Comment (cgrindel) [reply xcodeproj/internal/xcode_schemes.bzl:163]: @brentleyjones In another review comment, I suggest renaming `target` to `execution_context`. Does that sound reasonable to you?

**Review by cgrindel — CHANGES_REQUESTED** [review]

File: xcodeproj/internal/xcode_schemes.bzl (hunk 6 above)
Comment (erikkerber) [reply xcodeproj/internal/xcode_schemes.bzl:163]: lol holy moly

**Comment by chiragramani** [comment]
Thank you everyone for looking into it and adding the suggestions - it all makes sense. I will address them soon!

**Comment by brentleyjones** [comment]
@chiragramani Re-request a review when this is ready. I'm going to unsubscribe from it for now to reduce my notification noise 😄.

File: xcodeproj/internal/xcode_schemes.bzl (hunk 8 above)
Comment (chiragramani) [reply xcodeproj/internal/xcode_schemes.bzl:48]: Sounds good, thanks!

File: xcodeproj/internal/xcode_schemes.bzl (hunk 5 above)
Comment (chiragramani) [reply xcodeproj/internal/xcode_schemes.bzl]: Yes, it can be None. Updated to keep the naming consistent with test action - "expand_variables_based_on"

File: xcodeproj/internal/xcode_schemes.bzl (hunk 8 above)
Comment (chiragramani) [reply xcodeproj/internal/xcode_schemes.bzl:48]: Yeah, that makes sense!

File: tools/generator/src/DTO/XcodeScheme.swift [hunk 12]
```diff
@@ -114,6 +118,33 @@ disabled, but the target is referenced in the scheme's \(keyPath.actionType) act
     }
 }
 
+// MARK: PrePostAction
+
+extension XcodeScheme {
+    struct PrePostAction: Equatable, Decodable {
+        let name: String
+        let expandVariablesBasedOn: VariableExpansionContext?
```
Comment (cgrindel) [inline tools/generator/src/DTO/XcodeScheme.swift:126]: This should not be optional. The `VariableExpansionContext` enum has a `none` value.

File: xcodeproj/internal/xcode_schemes.bzl [hunk 13]
```diff
@@ -6,6 +6,17 @@ load(":xcode_schemes_internal.bzl", "xcode_schemes_internal")
 
 _DEFAULT_BUILD_CONFIGURATION_NAME = "Debug"
 
+def _focus_actions(build_targets, actions):
+    # TODO:
```
Comment (cgrindel) [inline xcodeproj/internal/xcode_schemes.bzl:10]: FYI. It looks like you have an errant TODO comment.

File: tools/generator/src/Generator/XCSchemeInfo+BuildActionInfo.swift [hunk 14]
```diff
@@ -54,17 +62,29 @@ extension XCSchemeInfo.BuildActionInfo {
...
+                      preActions: buildAction.preActions.resolve(
+                          targetResolver: targetResolver,
+                          targetIDsByLabel: targetIDsByLabel,
+                          context: context
+                      ),
+                      postActions: buildAction.postActions.resolve(
+                          targetResolver: targetResolver,
+                          targetIDsByLabel: targetIDsByLabel,
+                          context: context
+                      ))
```
Comment (cgrindel) [inline tools/generator/src/Generator/XCSchemeInfo+BuildActionInfo.swift:88]: nit: In this repo, function calls that extend to multiple lines put all of the arguments on separate lines. I think that this should look like this:

```swift
        try self.init(
            targets: buildTargetInfos,
            preActions: buildAction.preActions.resolve(
                targetResolver: targetResolver,
                targetIDsByLabel: targetIDsByLabel,
                context: context
            ),
            postActions: buildAction.postActions.resolve(
                targetResolver: targetResolver,
                targetIDsByLabel: targetIDsByLabel,
                context: context
            ),
        )
```

**Review by cgrindel — CHANGES_REQUESTED** [review]
Looking good! When you have added tests, I'll be glad to take another look.

File: tools/generator/src/DTO/XcodeScheme.swift (hunk 12 above)
Comment (chiragramani) [reply tools/generator/src/DTO/XcodeScheme.swift:126]: Hey! I made this optional because the settings provider(or target) can be passed as None. This is also consistent with the type that the `TestAction` has for the same reason. 

In case we want to make it a non-optional, the bazel side of it has to convert None to "none" string OR we would have to come up with a custom decodable that handles the case when Bazel passes None(serialized to "null") but we expect it to be a non-optional.

I am more inclined in keeping it as an optional above (`VariableExpansionContext`) as the intent is pretty clear when None is passed (without we converting None to "none" and other custom decodable workarounds). PrePostActionInfo will have a non-optional `VariableExpansionContextInfo` though.

Let me know your thoughts and I can change the implementation accordingly.

File: xcodeproj/internal/xcode_schemes.bzl (hunk 13 above)
Comment (chiragramani) [reply xcodeproj/internal/xcode_schemes.bzl:10]: Yes, @brentleyjones mentioned that he will be helping with this. (https://github.com/buildbuddy-io/rules_xcodeproj/pull/1047/files#r964738570)

File: tools/generator/src/DTO/XcodeScheme.swift (hunk 12 above)
Comment (cgrindel) [reply tools/generator/src/DTO/XcodeScheme.swift:126]: Gotcha. I thought that we provided a value in Starlark, if `None` was specified. Changing that behavior is definitely outside the scope of your PR. Thanks.

**Review by cgrindel — APPROVED** [review]
Looks good. Nice work!

File: tools/generator/src/Generator/XCSchemeInfo+BuildActionInfo.swift [hunk 15]
```diff
@@ -38,7 +44,9 @@ extension XCSchemeInfo.BuildActionInfo {
                     ),
                     buildFor: buildTarget.buildFor
                 )
-            }
+            },
+            preActions: try original.preActions.resolveHosts(topLevelTargetInfos: topLevelTargetInfos),
+            postActions: try original.postActions.resolveHosts(topLevelTargetInfos: topLevelTargetInfos)
```
Comment (brentleyjones) [inline tools/generator/src/Generator/XCSchemeInfo+BuildActionInfo.swift:49]: nit: 80 chars:
```suggestion
            preActions: try original.preActions
                .resolveHosts(topLevelTargetInfos: topLevelTargetInfos),
            postActions: try original.postActions
                .resolveHosts(topLevelTargetInfos: topLevelTargetInfos)
```

File: tools/generator/src/Generator/XCSchemeInfo+BuildActionInfo.swift [hunk 16]
```diff
@@ -54,17 +62,31 @@ extension XCSchemeInfo.BuildActionInfo {
...
                 )
             }
-        try self.init(targets: buildTargetInfos)
+
+        try self.init(
+            targets: buildTargetInfos,
+            preActions: buildAction.preActions.prePostActionInfos(
+                targetResolver: targetResolver,
+                targetIDsByLabel: targetIDsByLabel,
+                context: context
```
Comment (brentleyjones) [inline tools/generator/src/Generator/XCSchemeInfo+BuildActionInfo.swift:83]: We should have a unique context for each of these. This one could be `` "creating a pre-action `PrePostActionInfo`" ``.

File: tools/generator/src/Generator/XCSchemeInfo+PrePostActionInfo.swift [hunk 17]
```diff
@@ -0,0 +1,62 @@
...
+}
+
+extension XCSchemeInfo.PrePostActionInfo {
+    init(
+        prePostAction: XcodeScheme.PrePostAction,
+        targetResolver: TargetResolver,
+        targetIDsByLabel: [BazelLabel: TargetID],
+        context: String
+    ) throws {
+        guard let originalTargetLabel = prePostAction.expandVariablesBasedOn?.targetLabel else {
```
Comment (brentleyjones) [inline tools/generator/src/Generator/XCSchemeInfo+PrePostActionInfo.swift:26]: nit: 80 chars:
```suggestion
        guard let originalTargetLabel =
            prePostAction.expandVariablesBasedOn?.targetLabel
        else {
```

File: tools/generator/src/Generator/XCSchemeInfo+PrePostActionInfo.swift [hunk 18]
```diff
@@ -0,0 +1,62 @@
...
+        self.init(name: prePostAction.name,
+                  expandVariablesBasedOn: expandVariablesBasedOn,
+                  script: prePostAction.script)
+    }
+}
+
+extension Sequence where Element == XCSchemeInfo.PrePostActionInfo {
+    func resolveHosts<TargetInfos: Sequence>(
+        topLevelTargetInfos: TargetInfos
+    ) throws -> [XCSchemeInfo.PrePostActionInfo] where TargetInfos.Element == XCSchemeInfo.TargetInfo {
```
Comment (brentleyjones) [inline tools/generator/src/Generator/XCSchemeInfo+PrePostActionInfo.swift:50]: nit: 80 chars:
```suggestion
    ) throws -> [XCSchemeInfo.PrePostActionInfo]
    where TargetInfos.Element == XCSchemeInfo.TargetInfo {
```

File: tools/generator/test/XCSchemeInfo+PrePostActionInfoTests.swift [hunk 19]
```diff
@@ -0,0 +1,303 @@
...
+
+final class XCSchemeInfoPrePostActionInfoTests: XCTestCase {
+    
+    let xcodeprojBazelLabel = BazelLabel("//foo")
+
+    lazy var xcodeScheme = try! XcodeScheme(
+        name: "My Scheme",
+        // swiftlint:disable:next force_try
+        buildAction: try! .init(targets: [.init(label: targetResolver.targets["A 2"]!.label),
+                                          .init(label: targetResolver.targets["W"]!.label)]),
```
Comment (brentleyjones) [inline tools/generator/test/XCSchemeInfo+PrePostActionInfoTests.swift:16]: nit:
```suggestion
        buildAction: try! .init(targets: [
            .init(label: targetResolver.targets["A 2"]!.label),
            .init(label: targetResolver.targets["W"]!.label),
        ]),
```

File: tools/generator/test/XCSchemeInfo+PrePostActionInfoTests.swift [hunk 20]
```diff
@@ -0,0 +1,303 @@
...
+
+        // when
+        let resolvedPrePostActionInfos = try [prePostActionInfo].resolveHosts(
+            topLevelTargetInfos: []
+        )
+
+        // then
+        resolvedPrePostActionInfos.forEach { actionInfo in
+            XCTAssertEqual(actionInfo.expandVariablesBasedOn.targetInfo?.hostResolution,
+                           XCSchemeInfo.HostResolution.none)
```
Comment (brentleyjones) [inline tools/generator/test/XCSchemeInfo+PrePostActionInfoTests.swift:300]: nit:
```suggestion
            XCTAssertEqual(
                actionInfo.expandVariablesBasedOn.targetInfo?.hostResolution,
                XCSchemeInfo.HostResolution.none
            )
```

File: tools/generator/test/XCSchemeInfo+VariableExpansionContextInfoTests.swift [hunk 21]
```diff
@@ -49,6 +49,24 @@ extension XCSchemeInfoVariableExpansionContextInfoTests {
     }
 }
 
+extension XCSchemeInfoVariableExpansionContextInfoTests {
+    func test_targetInfo() {
+        // given
+        let contextInfo = XCSchemeInfo.VariableExpansionContextInfo.target(unitTestTargetInfo)
```
Comment (brentleyjones) [inline tools/generator/test/XCSchemeInfo+VariableExpansionContextInfoTests.swift:55]: nit: 80 chars:
```suggestion
        let contextInfo = XCSchemeInfo.VariableExpansionContextInfo
            .target(unitTestTargetInfo)
```

File: xcodeproj/internal/xcode_schemes.bzl [hunk 22]
```diff
@@ -147,6 +162,23 @@ def unfocus_schemes(schemes, unfocused_targets):
 
     return focused_schemes
 
+def _pre_post_action(script, expand_variables_based_on, name = "Run Script"):
```
Comment (brentleyjones) [inline xcodeproj/internal/xcode_schemes.bzl:165]:
```suggestion
def _pre_post_action(*, name = "Run Script", script, expand_variables_based_on):
```

File: xcodeproj/internal/xcode_schemes.bzl [hunk 23]
```diff
@@ -147,6 +162,23 @@ def unfocus_schemes(schemes, unfocused_targets):
...
+    """Constructs a "Pre | Post action" for the "Build | Run | Test | Profile | Archive" step of the Scheme.
+
+    Args:
+        script: script text.
+        expand_variables_based_on: Optional. One of the specified target labels .
+        name: name of the action.
```
Comment (brentleyjones) [inline xcodeproj/internal/xcode_schemes.bzl:171]:
```suggestion
        name: Title of the script.
        script: The script text.
        expand_variables_based_on: Optional. The label of the target that 
            environment variables will expand based on.
```

File: xcodeproj/internal/xcode_schemes.bzl [hunk 24]
```diff
@@ -147,6 +162,23 @@ def unfocus_schemes(schemes, unfocused_targets):
...
+
+    Returns:
+        A `struct` representing a Scheme's step "Pre | Post" action.
```
Comment (brentleyjones) [inline xcodeproj/internal/xcode_schemes.bzl:174]:
```suggestion
        A `struct` representing a scheme's step pre or post action.
```

File: xcodeproj/internal/xcode_schemes.bzl [hunk 25]
```diff
@@ -147,6 +162,23 @@ def unfocus_schemes(schemes, unfocused_targets):
 
     return focused_schemes
 
+def _pre_post_action(script, expand_variables_based_on, name = "Run Script"):
+    """Constructs a "Pre | Post action" for the "Build | Run | Test | Profile | Archive" step of the Scheme.
```
Comment (brentleyjones) [inline xcodeproj/internal/xcode_schemes.bzl:166]:
```suggestion
    """Constructs a pre or post action for a step of the scheme.
```

File: xcodeproj/internal/xcode_schemes.bzl [hunk 26]
```diff
@@ -273,6 +322,8 @@ def make_xcode_schemes(bazel_labels):
         test_action = _test_action,
         DEFAULT_BUILD_CONFIGURATION_NAME = _DEFAULT_BUILD_CONFIGURATION_NAME,
         BUILD_FOR_ALL_ENABLED = xcode_schemes_internal.BUILD_FOR_ALL_ENABLED,
+        pre_action = _pre_post_action,
+        post_action = _pre_post_action,
```
Comment (brentleyjones) [inline xcodeproj/internal/xcode_schemes.bzl:326]:
```suggestion
        pre_action = _pre_post_action,
        post_action = _pre_post_action,
        DEFAULT_BUILD_CONFIGURATION_NAME = _DEFAULT_BUILD_CONFIGURATION_NAME,
        BUILD_FOR_ALL_ENABLED = xcode_schemes_internal.BUILD_FOR_ALL_ENABLED,
```

**Review by brentleyjones — COMMENTED** [review]
Mostly styling.

I'll push the focused targets stuff before merging.

@cgrindel Can you do one final round of logic review?

**Review by brentleyjones — DISMISSED** [review]

File: xcodeproj/internal/xcode_schemes_internal.bzl [hunk 27]
```diff
@@ -28,18 +28,24 @@ def _scheme(
...
+def _build_action(targets, pre_actions, post_actions):
     """Constructs a build action for an Xcode scheme.
 
     Args:
         targets: A `sequence` of `struct` values as created by
             `xcode_schemes.build_target`.
+        pre_actions: A `sequence` of `struct` values as created by
+            `xcode_schemes.pre_action`.
+        post_actions: A `sequence` of `struct` values as created by
+            `xcode_schemes.post_action`.
```
Comment (brentleyjones) [inline xcodeproj/internal/xcode_schemes_internal.bzl:40]:
```suggestion
            `xcode_schemes.pre_post_action`.
        post_actions: A `sequence` of `struct` values as created by
            `xcode_schemes.pre_post_action`.
```

**Review by brentleyjones — COMMENTED** [review]
And one more, let's just expose the single `pre_post_action` function.

**Review by cgrindel — APPROVED** [review]
LGTM

File: tools/generator/src/Generator/XCSchemeInfo+BuildActionInfo.swift (hunk 16 above)
Comment (chiragramani) [reply tools/generator/src/Generator/XCSchemeInfo+BuildActionInfo.swift:85]: Makes sense, I will add the unique contexts.

**Review by brentleyjones — APPROVED** [review]

**Review by brentleyjones — APPROVED** [review]

**Comment by brentleyjones** [comment]
Thanks again @chiragramani!

**Comment by chiragramani** [comment]
Thanks a lot, @brentleyjones and @cgrindel for all the help with this PR 🙌🏼
